gtest-parallel

```diff
@@ -18,13 +18,62 @@ import gzip
 import multiprocessing
 import optparse
 import os
+import signal
 import subprocess
 import sys
 import tempfile
+import thread
 import threading
 import time
 import zlib
 
+# An object that catches SIGINT sent to the Python process and notices
+# if processes passed to wait_for_or_terminate() die by SIGINT (we
+# need to look for both of those cases, because pressing Ctrl+C can
+# result in either the main process or one of the subprocesses getting
+# the signal).
+#
+# Before a SIGINT is seen, wait_for_or_terminate(p) will simply call
+# p.wait() and return the result. Once a SIGINT has been seen (in the
+# main process or a subprocess, including the one the current call is
+# waiting for), wait_for_or_terminate(p) will call p.terminate() and
+# raise ProcessWasInterrupted.
+class SigintHandler(object):
+  class ProcessWasInterrupted(Exception): pass
+  def __init__(self):
+    self.__lock = threading.Lock()
+    self.__processes = set()
+    self.__got_sigint = False
+    signal.signal(signal.SIGINT, self.__sigint_handler)
+  def __on_sigint(self):
+    self.__got_sigint = True
+    while self.__processes:
+      try:
+        self.__processes.pop().terminate()
+      except OSError:
+        pass
+  def __sigint_handler(self, signal_num, frame):
+    with self.__lock:
+      self.__on_sigint()
+  def got_sigint(self):
+    with self.__lock:
+      return self.__got_sigint
+  def wait_for_or_terminate(self, p):
+    with self.__lock:
+      if self.__got_sigint:
+        p.terminate()
+        raise self.ProcessWasInterrupted
```

Comment:
Maybe add a comment instead of this line though, waiting on the terminated process should be fine.